lib/plugins/cors.js

```diff
@@ -27,27 +29,72 @@ var EXPOSE_HEADERS = [
     'response-time'
 ];
 
-// Normal
-var AC_ALLOW_ORIGIN = 'Access-Control-Allow-Origin';
-var AC_ALLOW_CREDS = 'Access-Control-Allow-Credentials';
-var AC_EXPOSE_HEADERS = 'Access-Control-Expose-Headers';
+var AC_REQ_METHOD = 'access-control-request-method';
+var AC_REQ_HEADERS = 'access-control-request-headers';
+var AC_ALLOW_CREDS = 'access-control-allow-credentials';
+var AC_ALLOW_ORIGIN = 'access-control-allow-origin';
+var AC_ALLOW_HEADERS = 'access-control-allow-headers';
+var AC_ALLOW_METHODS = 'access-control-allow-methods';
+var AC_EXPOSE_HEADERS = 'access-control-expose-headers';
+var AC_MAX_AGE = 'access-control-max-age';
+var AC_MAX_AGE_MS = 3600;
+var STR_VARY = 'vary';
+var STR_ORIGIN = 'origin';
 
 
 ///--- Internal Functions
 
-function matchOrigin(req, origins) {
-    var origin = req.headers.origin;
 
-    function belongs(o) {
-        if (origin === o || o === '*') {
-            origin = o;
-            return (true);
-        }
+/**
+ * turn potential string patterns into regexps.
+ * @private
+ * @function toRegExp
+ * @param {String} str an input string
+ * @returns {RegExp}
+ */
+function toRegExp(str) {
 
-        return (false);
+    if (str instanceof RegExp || str === '*') {
+        return str;
     }
 
-    return ((origin && origins.some(belongs)) ? origin : false);
+    var regexStr = str
+        // escape all the special characters, except *
+        .replace(/[\-\[\]\/\{\}\(\)\+\?\.\\\^\$\|]/g, '\\$&')
+        // replace all instances of * with a proper regex wildcard
+        .replace(/\*/g, '.*');
+
+    return new RegExp('^(https?://)?' + regexStr + '$', 'i');
```

Comment:
Should we be pre-pending the protocol? CORS is domain specific no?

Perhaps it would be better to scrub the protocol from `origin` before doing the test, i.e. `url.parse(origin).hostname`.